lib/async.js

```diff
@@ -1186,10 +1186,18 @@
             var result;
             try {
                 result = func.apply(this, args);
+                // if result is Promise object
+                if (typeof result.then === "function") {
+                    result.then(function(values) {
+                        var args = [null].concat(values);
+                        callback.apply(this, args);
+                    }).catch(callback);
+                } else {
+                    callback(null, result);
```

Comment:
Yep: The failing test is the one that guards against this: https://travis-ci.org/caolan/async/jobs/70115957